### ISSUE TITLE
Add Cedric to active contributors team

### ISSUE THREAD

**Issue by SimonBiggs** [pymedphys/pymedphys#1579]
Hi @cedricbelanger,

Would you be interested in adding yourself to the list of active contributors?

https://github.com/pymedphys/pymedphys/blob/main/lib/pymedphys/docs/README.rst

You do it by editing the above file, then after you've made the edit you need to run:

```
poetry run pymedphys dev propagate
```

As that readme text needs to be copied to a few locations.

**Comment by SimonBiggs** [comment]
Also, feel free to drop a logo into the following directory:

https://github.com/pymedphys/pymedphys/tree/main/logos

Just make sure that it is 200px x 200px (can have a rectangle image, just need to buffer it with white space to make it 200x200). Then when linking the image within the readme, make sure to use the commit hash within the image link.

**Comment by cedricbelanger** [comment]
Hi @SimonBiggs,

Yes I would be glad to join, thanks for the invitation ! :) I will add the info by the end of the week. 

I wonder, what is the ultimate aim of this repo? I just finished a project on commissioning our in-house TPS for HDR brachy (python based for the GUI part). I developed a fast TG43 calculation algorithm on CPU and GPU (C++ and CUDA), which is validated. With this we could easily implement 3D dose and DVH calculations from RTPLAN, which could be usefull. Would you be interested in adding such things in this projet in the future? Also, I developed a routine to calculate DSC metrics and hausdorff distance (that could be more efficient) for 3D dose comparisons as well to extend the gamma analysis. This might be useful for the community as well. 

Best,

Cédric

**Comment by SimonBiggs** [comment]
> I wonder, what is the ultimate aim of this repo?

To facilitated Medical Physicists sharing and reviewing one another's code. Specifically things that are directly practical for Medical Physics research or have direct utility within the Medical Physicist's day-to-day tasks within the clinic. 

To achieve this, the repository itself has a range of the "hard things" already set up, such as automated-testing, reproducible software distribution, hosted documentation, etc.

For the review side, the goal is to build a community of Medical Physicists who are interested open source software development and provide an easy means + framework for them to review one another's software.

---

> Would you be interested in adding such things in this projet in the future?

This stuff is awesome. The PyMedPhys repo and its community is built around enthusiastic contributors, so this sounds awesome :). I do have a range of caveat though which I've detailed below. They're not deal breakers, but just a bit of digging into the gory details of what would be involved / required for those particular projects you mentioned.

---

The key is, you would need to be willing to be a "code owner" of these projects. Essentially taking on the responsibility to review code submissions that people make to that portion of the code base, as well as be willing to field questions on the discourse forum / GitHub issues that are related to that portion of the code.

It certainly not a "set-and-forget" upload. It's more a "integrate and continuously tend" agreement. So only upload code that you believe you would be happy maintaining across its long term lifetime.

> With this we could easily implement 3D dose and DVH calculations from RTPLAN, which could be useful.

Yup, I agree that would be awesome. Certainly if it's something you find that you need to / want to be able to regularly do.

> I developed a fast TG43 calculation algorithm on CPU and GPU (C++ and CUDA), which is validated.

I personally do not have the experience to be reviewing raw CUDA code, and my C++ coding skills are rather limited. We also don't currently have the C++ testing, building, and release infrastructure developed around PyMedPhys at the moment, so in order to contribute this you would need to also be building that infrastructure into the PyMedPhys repo, including a means to build the resulting PyMedPhys wheel for each OS and architecture for new releases (and then committing to maintain this built infrastructure in the long term).

@hdclark has some done some amazing work over at https://github.com/hdclark/DICOMautomaton. And his stack is almost entirely C++, so he would have a lot of that infrastructure within his repository already. One option that was discussed in the past was to have something like a Python [SWIG](http://www.swig.org/) wrapper around @hdclark's work done in such a way that PyMedPhys can then have his work as an optional dependency within PyMedPhys to enable all sorts of awesomeness. An approach similar to that would keep the C++ build infrastructure out of PyMedPhys itself but still enable users of PyMedPhys to access the underlying tools using a Python API.

As another approach, have you seen JAX? https://jax.readthedocs.io/en/latest/index.html

It is an amazing library which allows for writing code in Python that then compiles to XLA which can be easily scheduled to run across multiple machines, on CPU, GPU, or even TPUs. That means we would write the same code and that very same code can run on a range of accelerators, and even be parallelised across a network of accelerators. Likely not an issue for TG43 calculations, but, being able to write one set of Python code and have the calculation be able to run on CPU + GPU + TPU is very appealing from a long term code maintenance point of view.

Here are some examples for JAX timings:
https://rlouf.github.io/post/jax-random-walk-metropolis/

I use JAX within my cloud infrastructure to speed up all sorts of tasks across multiple GPUs on multiple nodes within a Google Cloud Kubernetes cluster. It is both awesome to use and a very powerful tool :slightly_smiling_face:.

Also, a benefit of using JAX, it is being spearheaded by Google using the same technology that underpins TensorFlow. So any future speed ups / CUDA version support / etc is all handled by Google and automatically fed through to libraries that use JAX when they upgrade.

> I developed a routine to calculate DSC metrics and hausdorff distance

How does it compare to DeepMind's implementation (https://github.com/deepmind/surface-distance)? There are a range of hausdorff and DSC implementations out there. We would want to make sure that there is an important differentiator between the new code and what is already within the open source ecosystem.

**Comment by SimonBiggs** [comment]
Hi @cedricbelanger,

We're looking to make a release tomorrow to include your fix. If you don't manage to get in the team addition by then that's okay, it'll just mean it won't show up with the release on pypi:

https://pypi.org/project/pymedphys/

**Comment by cedricbelanger** [comment]
Hi @SimonBiggs Thanks for the recall. I've been insanely busy the past weeks with university. I don't know if I'll be able to contribute more to this project so I prefer not to integrate the contributor team yet. Anyways, I'll keep using this repo and share it with my collegues :).

Best regards,

Cédric

**Comment by SimonBiggs** [comment]
Kk, your DICOM co-ordinate fix has been massively appreciated. Thank you
for the work you put in to making it happen :)

We are aiming to have a 0.38.0 release with this fix either today or next
week.

Cheers,
Simon

**Comment by hdclark** [comment]
Hi @cedricbelanger 

I'm a bit late here, but what you mentioned above sounds interesting. Is the code posted anywhere yet?

I've tried to avoid CUDA as much as possible for portability reasons, but depending on the architecture it might be possible to conditionally compile when available and fallback on CPU/SIMD, or simply not bundle on CUDA-less systems. I do this already for a large geometry library that doesn't cross-compile easily.

RTPLAN handling and DVH extraction from (DICOM or other) 3D dose already exists in DICOMautomaton. A brachy dose engine does not, so this would be a welcomed addition. Iteration on the gamma analysis via Hausdorf metric would also be welcome!
